TiFlash crash in ~MemoryTrackerSetter()

Commit hash: a46271dd7db654ec0ab59f77915706863bf91e0b
Triggered when random_cop_send_failure_failpoint is triggered
BaseDaemon info:
[2023/12/18 13:15:50.881 +08:00] [ERROR] [BaseDaemon.cpp:367] [########################################] [source=BaseDaemon] [thread_id=14749]
[2023/12/18 13:15:50.881 +08:00] [ERROR] [BaseDaemon.cpp:368] ["(from thread 20) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=14749]
[2023/12/18 13:15:50.881 +08:00] [ERROR] [BaseDaemon.cpp:396] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=14749]
[2023/12/18 13:15:50.882 +08:00] [ERROR] [BaseDaemon.cpp:404] ["Access: read."] [source=BaseDaemon] [thread_id=14749]       
[2023/12/18 13:15:50.882 +08:00] [ERROR] [BaseDaemon.cpp:416] ["Unknown si_code."] [source=BaseDaemon] [thread_id=14749]
       0xb2543adfaultSignalHandler(int, siginfo_t*, void*) [tiflash+186991533]
                libs/libdaemon/src/BaseDaemon.cpp:211
  0x7f833e00bdb0<unknown symbol> [libc.so.6+347568]
       0x474c919long std::__1::__cxx_atomic_fetch_sub<long>(std::__1::__cxx_atomic_base_impl<long>*, long, std::__1::memory_order) [tiflash+74762521]
                /usr/local/bin/../include/c++/v1/atomic:1082
       0x474b0a3std::__1::__atomic_base<long, true>::fetch_sub(long, std::__1::memory_order) [tiflash+74756259]
                /usr/local/bin/../include/c++/v1/atomic:1736
       0x4759307MemoryTracker::free(long) [tiflash+74814215]
                dbms/src/Common/MemoryTracker.cpp:239
       0x4759381MemoryTracker::free(long) [tiflash+74814337]
                dbms/src/Common/MemoryTracker.cpp:256
       0x475bd20CurrentMemoryTracker::submitLocalDeltaMemory() [tiflash+74824992]
                dbms/src/Common/MemoryTracker.cpp:375
       0x474c383DB::MemoryTrackerSetter::~MemoryTrackerSetter() [tiflash+74761091]
                dbms/src/Common/MemoryTrackerSetter.h:41
       0xb1d752bDB::DM::MergedUnit::setFinish() [tiflash+186479915]
                dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h:47
       0xb1e341cDB::DM::MergedTask::setUnitFinish(int) [tiflash+186528796]
                dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h:129
       0xb1e3344DB::DM::MergedTask::readOneBlock() [tiflash+186528580]
                dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp:97
       0xb1e304eDB::DM::MergedTask::readBlock() [tiflash+186527822]
                dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp:32
       0xb1ed7d0DB::DM::SegmentReader::readSegments() [tiflash+186570704]
                dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:93
       0xb1ece35DB::DM::SegmentReader::run() [tiflash+186568245]
                dbms/src/Storages/DeltaMerge/ReadThread/SegmentReader.cpp:148
       0xb1ef511decltype(*(static_cast<DB::DM::SegmentReader*>(fp0)).*fp()) std::__1::__invoke<void (DB::DM::SegmentReader::*)(), DB::DM::SegmentReader*, void>(void (DB::DM::SegmentReader::*&&)(), DB::DM::SegmentReader*&&) [tiflash+186578193]
                /usr/local/bin/../include/c++/v1/type_traits:3859
       0xb1ef4bdvoid std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::DM::SegmentReader::*)(), DB::DM::SegmentReader*, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::DM::SegmentReader::*)(), DB::DM::SegmentReader*>&, std::__1::__tuple_indices<2ul>) [tiflash+186578109]
                /usr/local/bin/../include/c++/v1/thread:280
       0xb1ef22bvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::DM::SegmentReader::*)(), DB::DM::SegmentReader*> >(void*) [tiflash+186577451]
                /usr/local/bin/../include/c++/v1/thread:291
  0x7f833e056802start_thread [libc.so.6+653314]

--- a/dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h
+++ b/dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h
@@ -38,10 +38,14 @@ struct MergedUnit
     {
         if (!isFinished())
         {
-            // For updating memory statistics of `MemoryTracker`.
-            MemoryTrackerSetter setter(true, pool->mem_tracker.get());
-            task = nullptr;
-            stream = nullptr;
+            {
+                // For updating memory statistics of `MemoryTracker`.
+                MemoryTrackerSetter setter(true, pool->mem_tracker.get());
+                task = nullptr;
+                stream = nullptr;
+            }
+            // `SegmentReadTaskScheduler` will release `pool` if there is no other component/thread hold it.
+            // So `pool` should release after `setter` destructed because it holds a raw pointer of `mem_tracker`.
             pool = nullptr;
             return true;
         }

